### DISCUSSION TITLE
Updating variable at the end of the time step

### DISCUSSION THREAD

**Discussion by Bala-1005** [idaholab/moose#16997] [Q&A Modules: Solid mechanics] [answered]
Hello everyone,

Is there a way to manually update a variable say, elastic strain to a value of zero at a particular region based on the order parameter at the end of every timestep?

Thanks,
Bala

**Comment by GiudGiud** [comment, marked as answer]
Hello

I think a ConstantAux(Kernel) does most of what you're looking for
(https://mooseframework.inl.gov/source/auxkernels/ConstantAux.html)

You mean the order parameter of the variable? You'll have to modify the AuxKernel for that. 
You can get the order with .order() from a variable, as shown here for a base class that your variable likely inherits from (https://mooseframework.inl.gov/docs/doxygen/moose/classMooseVariableBase.html)

Best,
Guillaume

**Reply by Bala-1005** [reply]
Thank you @GiudGiud . I will try out the options. 
Essentially, I need to bring elastic strain to zero(at the end of each timestep) at parts of the domain where the phase field variable is 0 in order to compensate for the lack of stiffness in those regions.

Thanks,
Bala

**Reply by hugary1995** [reply]
Are your displacement variables nonlinear? If so this won't work.
You'll have to modify the Strain/Stress calculator to achieve that.

**Reply by hugary1995** [reply]
Let me clarify. There are two types of variables in MOOSE, `Variables` and `AuxVariables`. The `Variables` are essentially solutions to your system, and `AuxVariables`, according to their names, are auxilliary, and you can set their values manually as you wish. You cannot run `AuxKernels` on `Variables`.

I guess you have displacement variables in your input files, and they are `Variables` instead of `AuxVariables`. In your input file, you might have a strain calculator that calculates a material property called "elastic_strain" or "mechanical_strain" based on the current displacements. What you want to do here, is to add a condition in your strain calculator such that, if the value of your coupled order parameter is above (or below) certain value, you set the `_elastic_strain[_qp] = 0`. There might be existing classes for this particular purpose.

**Reply by Bala-1005** [reply]
Thank you for the explanation @hugary1995 . This is exactly what I want. Yes, my displacements are variables.

**Reply by Bala-1005** [reply]
I have been trying to write an if condition such that the body of the if condition executes at the end of timestep. How do I access exec_timestep_end? Or is there any other way to make a custom user material access the end of timestep? 

Thanks,
Bala

**Reply by GiudGiud** [reply]
sorry we missed this.
You can get the current execute on flag from the fe_problem with the getCurrentExecuteOnFlag routine
